server/functions/src/refreshcasestats.ts

```diff
@@ -0,0 +1,371 @@
+//WHO_CASE_STATS_URL = "http://localhost:8000/arcgisdataElided.txt";
+//WHO_CASE_STATS_URL = "http://localhost:8000/x.txt";
+
+// Peek or push last entry from timeseries.
+function getSnapshotForTimestamp(data: who.CaseStats, timestamp: number) {
+  if (data.timeseries.length != 0) {
+    let snapshot = data.timeseries[data.timeseries.length - 1];
+    if (snapshot.epochMsec > timestamp) {
+      throw new Error(
+        `Unsorted country data found at ${timestamp} for {data.jurisdictionType} {data.jurisdiction}.`
+      );
+    }
+    if (snapshot.epochMsec == timestamp) {
+      if (data.jurisdictionType != who.JurisdictionType.GLOBAL) {
+        // For individual jurisdictions, we should NOT see the same timestamp
+        // more than once.
+        // Fail and hope that the next run of the cron job succeeds.
+        // Note that when the bad results are doubled/trebled/etc, they have
+        // also been missing data, so ignoring duplicates is not safe.
+        throw new Error(
+          `Duplicate country data found at ${timestamp} for {data.jurisdictionType} {data.jurisdiction}.`
+        );
+      }
+      return snapshot;
+    }
+  }
+
+  let snapshot = {
+    epochMsec: timestamp,
+
+    dailyCases: 0,
+    dailyDeaths: 0,
+    dailyRecoveries: -1,
+
+    totalCases: 0,
+    totalDeaths: 0,
+    totalRecoveries: -1,
+  } as who.StatSnapshot;
+
+  data.timeseries.push(snapshot);
+  return snapshot;
+}
+
+function updateData(
+  data: who.CaseStats,
+  timestamp: number,
+  dailyDeaths: number,
+  totalDeaths: number,
+  dailyCases: number,
+  totalCases: number
+) {
+  if (timestamp > data.lastUpdated) {
+    data.lastUpdated = timestamp;
+  }
+  // Note that data.totalCases is a sum of dailyCases,
+  // while snapshot.totalCases is simply copied from the input data.
+  // The totalCases at the most recent snapshot should match
+  // data.totalCases, but this system does not verify that.
+  data.cases += dailyCases;
+  data.deaths += dailyDeaths;
+
+  let snapshot = getSnapshotForTimestamp(data, timestamp);
+
+  snapshot.dailyCases += dailyCases;
+  snapshot.dailyDeaths += dailyDeaths;
+  snapshot.totalCases += totalCases;
+  snapshot.totalDeaths += totalDeaths;
+}
+
+/**
+ * Process the JSON data returned by the ArcGIS server.
+ *
+ * @param features        JSON features from ArcGIS server
+ * @param countryData Accumulated data for each country.
+ * @param globalData  Accumulated data, sum of all countries.
+ */
```

Comment:
Comments like this are best placed directly above the function definition.

Reply:
Hah. Also, what is the ts docstring format anyway? This is javadoc.